hw/formal/tools/jaspergold/parse-fpv-report.py

```diff
@@ -0,0 +1,217 @@
+    '''Count errors, warnings, and property status from the log file'''
+    message_patterns = [("errors", r"^ERROR: .*"),
+                        ("errors", r"^\[ERROR.*"),
+                        ("warnings", r"^WARNING: .*"),
+                        ("warnings", r"^\[WARN.*"),
+                        ("proven", r"^\[\d+\].*proven.*"),
+                        ("cex", r"^\[\d+\].*cex.*"),
+                        ("covered", r"^\[\d+\].*covered.*"),
+                        ("undetermined", r"^\[\d+\].*undetermined.*"),
+                        ("unreachable", r"^\[\d+\].*unreachable.*")]
+    fpv_summary = extract_messages_count(str_buffer, message_patterns)
+
+    total_msg = fpv_summary["proven"] + fpv_summary["cex"] + fpv_summary["undetermined"]
+    if total_msg > 0:
+        pass_rate = (fpv_summary["proven"] + fpv_summary["undetermined"]) / total_msg * 100
+    else:
+        pass_rate = 0
+    fpv_summary["pass_rate"] = "{0:.2f} %".format(round(pass_rate, 2))
+
+    total_msg = fpv_summary["covered"] + fpv_summary["unreachable"]
+    if total_msg > 0:
+        cov_rate = fpv_summary["covered"] / total_msg * 100
+    else:
+        cov_rate = 0
+    fpv_summary["cov_rate"] = "{0:.2f} %".format(round(cov_rate, 2))
+
+    return fpv_summary
+
+
+def get_results(logpath):
+    '''Parse FPV log file and extract info to a dictionary'''
+    results = OrderedDict()
+    fpv_messages = OrderedDict()
+    fpv_summary = OrderedDict()
+    try:
+        with Path(logpath).open() as f:
+            full_file = f.read()
+            fpv_messages = parse_fpv_message(full_file)
+            fpv_summary = parse_fpv_summary(full_file)
+
+    except IOError as err:
+        log.error("[get_results] IOError %s", err)
+        fpv_messages['errors'].append('IOError: %s' % err)
+
+    results["messages"] = fpv_messages
+    if fpv_summary:
+        results["fpv_summary"] = fpv_summary
+
+    return results
+
+
+def get_cov_results(logpath):
+    '''Parse FPV coverage information from the log file'''
+    cov_results = OrderedDict()
+    try:
+        with Path(logpath).open() as f:
+            full_file = f.read()
+            coverage_patterns = [("stimuli", r"^\|\w+.*\|(\d*.\d\d)%.*\|\d*.\d\d%.*\|\d*.\d\d%.*"),
+                                 ("coi", r"^\|\w+.*\|\d*.\d\d%.*\|(\d*.\d\d)%.*\|\d*.\d\d%.*"),
+                                 ("proof", r"^\|\w+.*\|\d*.\d\d%.*\|\d*.\d\d%.*\|(\d*.\d\d)%.*")]
+            cov_results.update(extract_messages(full_file, coverage_patterns))
+            for key, item in cov_results.items():
+                if len(item) == 1:
+                    cov_results[key] = item[0] + " %"
+                else:
+                    cov_results[key] = "N/A"
+                    log.error("Parse %s coverage error. Expect one matching value, get %s",
+                              key, item)
+
+    except IOError as err:
+        log.error("[get_cov_results] IOError %s", err)
+        cov_results = None
```

Comment:
Probably better to return None here and return `cov_results` from the try block. Then `cov_results` doesn't need defining and updating: you can just set it equal to the result of `extract_messages`. Something like this:
```python
def get_cov_results(logpath):
    '''Parse FPV coverage information from the log file'''
    try:
        with Path(logpath).open() as f:
            full_file = f.read()
            coverage_patterns = [("stimuli", r"^\|\w+.*\|(\d*.\d\d)%.*\|\d*.\d\d%.*\|\d*.\d\d%.*"),
                                 ("coi", r"^\|\w+.*\|\d*.\d\d%.*\|(\d*.\d\d)%.*\|\d*.\d\d%.*"),
                                 ("proof", r"^\|\w+.*\|\d*.\d\d%.*\|\d*.\d\d%.*\|(\d*.\d\d)%.*")]
            cov_results = extract_messages(full_file, coverage_patterns)
            for key, item in cov_results.items():
                if len(item) == 1:
                    cov_results[key] = item[0] + " %"
                else:
                    cov_results[key] = "N/A"
                    log.error("Parse %s coverage error. Expect one matching value, get %s",
                              key, item)
            return cov_results

    except IOError as err:
        log.error("[get_cov_results] IOError %s", err)
        return None
```